[CORL-810] Custom Moderation Phases

wyattjoh opened the pull request:
<!--

Thank you for submitting a pull request! Please note that by contributing to
Coral, you agree to our Code of Conduct: http://code-of-conduct.voxmedia.com/

Before submitting your Pull Request (or PR), please verify that:

* [ ] Your code is up-to-date with the base branch
* [ ] You've successfully run `npm run test` locally

Refer to CONTRIBUTING.MD for more details.

  https://github.com/coralproject/talk/blob/master/CONTRIBUTING.md

-->

## What does this PR do?

Adds support for custom moderation phases to allow integration with external services to assist with implementing business rules via automatic moderation.

<!--

In this section, you should be describing what other Github issues or tickets
that this PR is designed to addressed.

Any related Github issue should be linked by adding its URL to this section.

-->

## What changes to the GraphQL/Database Schema does this PR introduce?

<!--

In this section, you should describe any changes to be made to the GraphQL
schema file (located https://github.com/coralproject/talk/blob/master/src/core/server/graph/schema/schema.graphql) or any
database model (located as types in the https://github.com/coralproject/talk/blob/master/src/core/server/models directory).

If no changes were added to the GraphQL/Database Schema as a part of this PR,
simply write "None".

-->

Introduces new types and renames some existing ones that have not been released yet. More details to come here.

- Added `Query.externalModerationPhase`
- Added `Mutation.createExternalModerationPhase`
- Added `Mutation.updateExternalModerationPhase`
- Added `Mutation.enableExternalModerationPhase`
- Added `Mutation.disableExternalModerationPhase`
- Added `Mutation.deleteExternalModerationPhase`
- Added `Mutation.rotateExternalModerationPhaseSigningSecret`

## How do I test this PR?

<!--

In this section, you should be describing any manual testing that can be used to
verify features introduced or bugs fixed in this PR.

 -->

TBD

wyattjoh committed:
feat: initial implementation

wyattjoh committed:
feat: renamed fields from mutations

wyattjoh committed:
fix: more renaming to streamline {Key,Secret}->SigningSecret

wyattjoh committed:
feat: introduced WrappedInternalError

wyattjoh committed:
feat: enhanced extern payload, more fetch options

- Added tenant.{id,domain} to extern payload
- Added site.id to the extern payload
- Added response size limit to fetch
- Added new SCRAPE_MAX_RESPONSE_SIZE env var for managing the size of
  responses for scraping

wyattjoh committed:
fix: fixed bug with scrape invocation

wyattjoh committed:
feat: added more queries + mutations

- Added Query.externalModerationPhase
- Added Mutation.createExternalModerationPhase
- Added Mutation.updateExternalModerationPhase
- Added Mutation.enableExternalModerationPhase
- Added Mutation.disableExternalModerationPhase
- Added Mutation.deleteExternalModerationPhase
- Added Mutation.rotateExternalModerationPhaseSigningSecret

wyattjoh committed:
feat: added secret management

wyattjoh committed:
fix: linting

wyattjoh committed:
fix: merge conflict fix

wyattjoh committed:
feat: added UI

wyattjoh committed:
fix: linting

wyattjoh committed:
fix: linting

wyattjoh committed:
fix: updated snapshots

wyattjoh committed:
fix: improved docs

wyattjoh committed:
fix: improved docs

wyattjoh committed:
fix: added locales

tessalt commented on src/core/server/models/settings/settings.ts:
```diff
@@ -85,6 +87,71 @@ export type Auth = Omit<GQLAuth, "integrations"> & {
+  createdAt: Date;
+}
+
+export interface ExternalExternalIntegration {
```
this naming is kind of wild but I get where you're coming from. what if we changed this to just `ExternalIntegration` and changed `ExtrenalIntegration` to `ThirdPartyIntegration` to differentiate the two types of external? or maybe "customer integration" vs "third-party integration"?

wyattjoh replied:
This feels like a renaming error, will take a look!

tessalt commented on src/core/server/stacks/createComment.ts:
```diff
@@ -91,7 +92,7 @@ const markCommentAsAnswered = async (
     story.settings.expertIDs.some((id) => id === author.id) &&
     // And this is the first reply (depth of 1)...
-    comment.ancestorIDs.length === 1
+    getDepth(comment) === 1
```
🆒

tessalt reviewed (approved):
there's a lot going on here but looks good as far as I can tell. Looking forward to seeing a demo!

wyattjoh committed:
review: improve naming

alinhle commented on src/core/client/admin/routeConfig.tsx:
```diff
@@ -120,13 +124,33 @@ export default makeRouteConfig(
+            <Route
+              path=":phaseID"
+              {...ConfigureExternalModerationPhaseRoute.routeConfig}
+            />
```
Is it possible to have this route inside the above path `configure/moderation/phases`?

alinhle commented on src/core/client/admin/routes/Configure/sections/ModerationPhases/ConfigureExternalModerationPhase/ExternalModerationPhaseDangerZone.tsx:
```diff
@@ -0,0 +1,185 @@
+      await deleteExternalModerationPhase({ id: phase.id });
+
+      // Send the user back to the webhook endpoints listing.
+      router.push(urls.admin.webhooks);
```
This should go back to moderation phase I believe :-)

alinhle commented on src/core/client/admin/routes/Configure/sections/ModerationPhases/ConfigureExternalModerationPhase/RotateSigningSecretModal.tsx:
```diff
@@ -0,0 +1,168 @@
+            </AppNotification>
+          </Localized>
+        );
+        window.scroll(0, 0);
```
Doesn't scroll – in Chrome at least. Changing this to `setTimeout(() => window.scroll(0, 0));` does the trick. Should be applied to the webhook implementation too.

alinhle commented on src/core/client/admin/routes/Configure/sections/ModerationPhases/ConfigureExternalModerationPhaseForm/ConfigureExternalModerationPhaseForm.tsx:
```diff
@@ -0,0 +1,235 @@
+          // The external moderation phase was defined, update it.
+          await update(values);
+        } else {
+          // The external moderation phase wasn't defined, created it.
```
Just a small typo `created` => `create`

alinhle commented on src/core/client/admin/routes/Configure/sections/ModerationPhases/ConfigureExternalModerationPhaseForm/ConfigureExternalModerationPhaseForm.tsx:
```diff
@@ -0,0 +1,235 @@
+              )}
+            </Field>
+            <Field
+              name="timeout"
```
Need line `parse={parseInteger}` here otherwise it fails the GQL valdiator

alinhle reviewed (changes requested):
Looking great! I'll try to test it tomorrow.

alinhle commented:
Tested and works great! 👍 

One question: Don't you think only giving max 24h to deploy the new key could be too short?
![image](https://user-images.githubusercontent.com/14221600/81944182-ade53b80-95fc-11ea-8072-f0345fa47112.png)

wyattjoh commented:
Because the key rotation option is initiated by the operator, it's something they control. I think the 24 hour delay is more than acceptable given that.

wyattjoh committed:
review: some review changes

- Switched /moderation/phase to /moderation/phases
- Fixed scrolling
- Fixed redirection

wyattjoh committed:
fix: added scroll timeout for webhooks
